### ISSUE TITLE
Graceful shutdown of signal handlers

### ISSUE THREAD

**Issue by ahopkins** [sanic-org/sanic#2369]
When signals are added, we should run some sort of a cleanup like this on shutdown to make sure any custom signals are run. This also includes a suggestion that we identify them with a name: `signal-XXXXX`.

```python
async def cleanup(app, _):
    for task in asyncio.all_tasks():
        if task.get_name().startswith("signal"):
          await task
```

**Comment by ChihweiLHBird** [comment]
Is this issue somehow related to #1874 since they are both about cleaning in shutdown? And how was #1874 going so for?

**Comment by ahopkins** [comment]
Yes, exactly. But I still want to name all of our signal tasks. I think that is a good idea.

**Comment by ChihweiLHBird** [comment]
Thanks. So, named background tasks clean shutdown was done. Here we would need to clean named signal tasks (asyncio tasks, not background tasks) as well. Is my understanding correct? Do we also need to clean unnamed background tasks in the clean shutdown?

**Comment by jonathanstaniforth** [comment]
Hi @ahopkins 👋 , I'm interested in taking this as my first issue! I have some ideas and questions which I will discuss below.

If I followed the flow correctly, I believe the task for signal handlers gets created in the `sanic.signals.SignalRouter.dispatch` method. In which case, a name could be assigned to the task like so:

`task = asyncio.get_running_loop().create_task(dispatch, name="signal")`

For the `XXXXX` part of the signal task's name, what were you thinking of passing here?

For the running of signal tasks on shutdown, I was thinking that a new method could be added to `sanic.app.Sanic` that's similar to the `sanic.app.Sanic.shutdown_tasks` method. This method would then be called in the `sanic.mixins.runner.RunnerMixin.stop` method to run the signal tasks on shutdown. Is this approach similar to what you were looking for?
